### ISSUE TITLE
[168] Proximal Policy Optimization Algorithms

### ISSUE THREAD

**Issue by long8v** [long8v/PTIR#187]
<img width="703" alt="image" src="https://github.com/user-attachments/assets/ad3ddfeb-d439-4c96-b841-2353465edc1a">

[paper](https://arxiv.org/pdf/1707.06347)

## TL;DR
- **I read this because.. :** 배경지식 차
- **task :** RL 
- **problem :** q-learning은 너무 불안정하고, trpo 는 상대적으로 복잡. data efficient하고 sclable한 arch?
- **idea :** KL divergence term 대신에 clipping을 사용하자. step size $\beta$를 adaptive하게 변동시키자 
- **input/output :** state -> action 
- **architecture :** MLP 
- **objective :** $L^{CLIP}(\theta) = \hat{\mathbb{E}}_t \left[ \min(r_t(\theta)\hat{A}_t, \text{clip}(r_t(\theta), 1 - \epsilon, 1 + \epsilon)\hat{A}_t) \right]$
- **baseline :** loss(no clipping, KL penalty), A2C, CEM, TRPO
- **data :** OpenAI Gym(Mujoco), human control task(Roboschool), Atari
- **evaluation :** 몰라
- **result :** 좋다
- **contribution :** 간단하고 직관적인 loss.. 사실 이 논문만 보면 직관적인 이해가 되어서 굳이 그 전 내용을 알았어야 하는 생각도.. ㅎㅎ
- **etc. :**

### preliminary 
- policy gradient method
<img width="232" alt="image" src="https://github.com/user-attachments/assets/764d9bf8-cacd-4c12-9437-9784439ede96">

policy network에 대한 gradient를 advantage에 가중하여 미분하는 형태
여기서 E는 sample을 뽑아서 그냥 평균 취하는 형태

<img width="272" alt="image" src="https://github.com/user-attachments/assets/b7cab011-36d4-423a-aa6b-79e3f1e849c5">

loss는 이러한 policy를 따르는 trajectory에 대한 advantage 가중합의 기대값이 최대가 되도록 하면 된다
그러나 large policy update가 될 경우 성능이 좋지 않다

- Trust Region methods
TRPO는 대리 함수를 사용하여 성능 개선이 보장되는 policy update를 할 수 있는 surrogate 함수를 증명했고(https://github.com/long8v/PTIR/issues/154) 이는 policy가 업데이트 될 때의 제약이 생겨서 결론적으로 아래와 같이 new, old policy network의 importance weight와 advantage가 곱해진 형태 
<img width="367" alt="image" src="https://github.com/user-attachments/assets/e1489535-f648-4625-aef6-f77c6c7a0b9d">

실제로는 제약이 아니라 penalty term을 추가함 여기서 $beta$는 하이퍼파라미터임. 
<img width="460" alt="image" src="https://github.com/user-attachments/assets/610aa2b5-8f85-4906-a5a7-26f12bc02638">

이 논문에서 말하는 TRPO의 문제점은 $\beta$가 하나로 골라질 수 없다 정도인듯.

### Clipped Surrogate Objective
<img width="367" alt="image" src="https://github.com/user-attachments/assets/a7a3deb0-4cdc-4eb7-ba81-feb095a2dc3e">

<img width="415" alt="image" src="https://github.com/user-attachments/assets/f04bf66e-ec62-45b1-92dd-503eae6cf1b6">

결론적으로 $r_t(\theta) = \frac{\phi_\theta(a_t | s_t)}{\phi_{\theta_{old}}(a_t | s_t)}$가 너무 크거나 작지 않도록 clip했다고 보면 됨 

<img width="703" alt="image" src="https://github.com/user-attachments/assets/f6307ca3-f047-4b01-befe-d245209c85d8">

이렇게 하니 policy network의 변동이 적었다고 함

<img width="724" alt="image" src="https://github.com/user-attachments/assets/c6337d1e-19bf-4381-9d65-d04f6c6c2eee">

#### Adaptive KL Penalty Coefficient

<img width="635" alt="image" src="https://github.com/user-attachments/assets/6b51ff1b-6408-4d9e-ab0a-1a298efa1919">

아이참 이것도 아주 직관적이고 간단하다.. 

### Experiments
- loss term
<img width="723" alt="image" src="https://github.com/user-attachments/assets/07cbe885-9056-4e59-ab1e-d74f0318112b">

-  other RL algorithms
봐도 모르쥬 ㅎ
<img width="716" alt="image" src="https://github.com/user-attachments/assets/3cb58942-489f-4be1-97f6-e00cfecb2fd3">

**Comment by long8v** [comment]
def learn
```python
                rollout = self.runner.run(callback)
                # Unpack
                obs, returns, masks, actions, values, neglogpacs, states, ep_infos, true_reward = rollout
```

def setup_models (loss 정의부)
https://github.com/Stable-Baselines-Team/stable-baselines/blob/master/stable_baselines/ppo2/ppo2.py#L176-L188
```python
                    vf_losses1 = tf.square(vpred - self.rewards_ph)
                    vf_losses2 = tf.square(vpred_clipped - self.rewards_ph)
                    self.vf_loss = .5 * tf.reduce_mean(tf.maximum(vf_losses1, vf_losses2))

                    ratio = tf.exp(self.old_neglog_pac_ph - neglogpac)
                    pg_losses = -self.advs_ph * ratio
                    pg_losses2 = -self.advs_ph * tf.clip_by_value(ratio, 1.0 - self.clip_range_ph, 1.0 +
                                                                  self.clip_range_ph)
                    self.pg_loss = tf.reduce_mean(tf.maximum(pg_losses, pg_losses2))
                    self.approxkl = .5 * tf.reduce_mean(tf.square(neglogpac - self.old_neglog_pac_ph))
                    self.clipfrac = tf.reduce_mean(tf.cast(tf.greater(tf.abs(ratio - 1.0),
                                                                      self.clip_range_ph), tf.float32))
                    loss = self.pg_loss - self.entropy * self.ent_coef + self.vf_loss * self.vf_coef
```

<img width="462" alt="image" src="https://github.com/user-attachments/assets/2e0ee1eb-413c-4c18-9957-f407f239b63d">

- `vf_losses` : $r_t(\theta)$ where  $r_t(\theta) = \frac{\phi_\theta(a_t | s_t)}{\phi_{\theta_{old}}(a_t | s_t)}$
- `self.advs_ph` : $\hat A_t$
